addons/stock_account/wizard/stock_valuation_layer_revaluation.py

```diff
@@ -66,53 +94,50 @@ def action_validate_revaluation(self):
+        layers_with_qty = self.env['stock.valuation.layer'].search([
             ('product_id', '=', product_id.id),
             ('remaining_qty', '>', 0),
             ('company_id', '=', self.company_id.id),
         ])
+        adjusted_layers = self.adjusted_layer_ids or layers_with_qty
+        qty_to_adjust = sum(adjusted_layers.mapped('remaining_qty'))
+        value_change_to_apply = self.added_value
+        unit_value_change = self.currency_id.round(self.added_value / qty_to_adjust)
+        # adjust all layers by the unit value change per unit, except the last layer which gets
+        # whatever is left. This avoids rounding issues e.g. $10 on 3 products => 3.33, 3.33, 3.34
+        for layer in adjusted_layers:
+            if layer != adjusted_layers[-1]:
+                value_change = unit_value_change * layer.remaining_qty
+                layer.remaining_value += value_change
+                value_change_to_apply -= value_change
+            else:
+                layer.remaining_value += value_change_to_apply
 
-        # Create a manual stock valuation layer
-        if self.reason:
-            description = _("Manual Stock Valuation: %s.", self.reason)
-        else:
-            description = _("Manual Stock Valuation: No Reason Given.")
-        if product_id.categ_id.property_cost_method == 'average':
-            description += _(
-                " Product cost updated from %(previous)s to %(new_cost)s.",
-                previous=product_id.standard_price,
-                new_cost=product_id.standard_price + self.added_value / self.current_quantity_svl
-            )
-        revaluation_svl_vals = {
+            if float_compare(layer.remaining_value, 0, precision_rounding=self.product_id.uom_id.rounding) < 0:
+                raise UserError(_('The value of a stock valuation layer cannot be negative. Landed cost could be used to correct a specific transfer.'))
+
+        description = _("Manual Stock Valuation: %s.", self.reason or _("No Reason Given."))
+        # Update the stardard price in case of AVCO/FIFO
+        cost_method = product_id.categ_id.property_cost_method
+        if cost_method in ['average', 'fifo']:
+            previous_cost = product_id.standard_price
+            total_product_qty = sum(layers_with_qty.mapped('remaining_qty'))
+            product_id.with_context(disable_auto_svl=True).standard_price += self.added_value / total_product_qty
+            if cost_method == 'average':
+                description += _(
+                    " Product cost updated from %(previous)s to %(new_cost)s.",
+                    previous=previous_cost,
+                    new_cost=product_id.standard_price
+                )
+
+        previous_value_svl = self.current_value_svl
+        revaluation_layer = self.env['stock.valuation.layer'].create({
             'company_id': self.company_id.id,
             'product_id': product_id.id,
             'description': description,
```

Comment:
I think it's a good idea to keep track of specific layer updated. Maybe add the ids or the reference in the description